pkg/streamingpromql/operators/aggregation.go

```diff
@@ -429,9 +443,80 @@ func (a *Aggregation) accumulateSeriesIntoGroup(s types.InstantVectorSeriesData,
 		seriesGroup.histogramSums = seriesGroup.histogramSums[:steps]
 	}
 
+	err = a.aggregationFunc(s, start, interval, seriesGroup)
+	if err != nil {
+		if err := functions.NativeHistogramErrorToAnnotation(err, a.emitAnnotationFunc); err != nil {
+			// Unknown error: we couldn't convert the error to an annotation. Give up.
+			return err
+		}
+	}
+
+	types.PutInstantVectorSeriesData(s, a.MemoryConsumptionTracker)
+	seriesGroup.remainingSeriesCount--
+	return nil
+}
+
+func (a *Aggregation) emitAnnotation(generator functions.AnnotationGenerator) {
+	metricName := a.metricNames.GetMetricNameForSeries(a.currentSeriesIndex)
+	a.Annotations.Add(generator(metricName, a.Inner.ExpressionPosition()))
+}
+
+func (a *Aggregation) Close() {
+	a.Inner.Close()
+}
+
+type groupSorter struct {
+	metadata []types.SeriesMetadata
+	groups   []*group
+}
+
+func (g groupSorter) Len() int {
+	return len(g.metadata)
+}
+
+func (g groupSorter) Less(i, j int) bool {
+	return g.groups[i].lastSeriesIndex < g.groups[j].lastSeriesIndex
+}
+
+func (g groupSorter) Swap(i, j int) {
+	g.metadata[i], g.metadata[j] = g.metadata[j], g.metadata[i]
+	g.groups[i], g.groups[j] = g.groups[j], g.groups[i]
+}
+
+type aggregationOperationFunc func(s types.InstantVectorSeriesData, start int64, interval int64, seriesGroup *group) error
+
+var aggregationFunctions = map[parser.ItemType]aggregationOperationFunc{
+	parser.MAX: maxAgg,
+	parser.SUM: sumAgg,
+}
+
+func maxAgg(s types.InstantVectorSeriesData, start int64, interval int64, seriesGroup *group) error {
 	for _, p := range s.Floats {
+		if math.IsNaN(p.F) {
+			continue
+		}
 		idx := (p.T - start) / interval
-		seriesGroup.floatSums[idx], seriesGroup.floatCompensatingMeans[idx] = floats.KahanSumInc(p.F, seriesGroup.floatSums[idx], seriesGroup.floatCompensatingMeans[idx])
+		if !seriesGroup.floatPresent[idx] || seriesGroup.floatPresent[idx] && p.F > seriesGroup.floatValues[idx] {
+			seriesGroup.floatValues[idx] = p.F
+			seriesGroup.floatPresent[idx] = true
+		}
+	}
+
+	// If a histogram exists max treats it as 0. We have to detect this here so that we return a 0 value instead of nothing.
```

Comment:
What happens if there's a histogram at T in one series in the group and a float less than 0 at T in another series in the group?

Reply:
the result would be the negative number (ie the float) in both engines. This happens because we just check for `floatPresent`. This comment therefore needs correcting.

Reply:
I think I was mistaken here. On re-testing it looks like promql always treats it as 0. So a histogram is bigger than a negative float.

Reply:
If there's a positive float value and a histogram, is the positive float value returned?

Reply:
Yes. This is confirmed in the [tests](https://github.com/grafana/mimir/pull/9008/files#diff-582a43a76c5a770fb1f0a8ca8b43847c58802fd9f4a9736d6639ebb18bd7993a).

Reply:
I'm thinking this sounds like a bug in prometheus actually. In that max(float, histogram) should always just return the float. We should keep this to be consistent though for now, but I'll open something upstream and link to it on the comment

Reply:
>I'm thinking this sounds like a bug in prometheus actually.

Yep, this behaviour sounds weird to me.

Reply:
https://github.com/prometheus/prometheus/issues/14711

I'm going to mark this discussion as resolved. We're being consistent with promql, and have opened an issue there.